tikv-client/src/main/java/com/pingcap/tikv/operation/iterator/ScanIterator.java

```diff
@@ -83,66 +76,45 @@ private boolean loadCache() {
       // Session should be single-threaded itself
       // so that we don't worry about conf change in the middle
       // of a transaction. Otherwise below code might lose data
-      if (currentCache.size() < batchSize) {
+      if (currentCache.size() < session.getConf().getScanBatchSize()) {
         // Current region done, start new batch from next region
         startKey = region.getEndKey();
-        if (startKey.size() == 0 || !contains(startKey)) {
-          return false;
-        }
       } else {
         // Start new scan from exact next key in current region
         Key lastKey = Key.toRawKey(currentCache.get(currentCache.size() - 1).getKey());
         startKey = lastKey.next().toByteString();
       }
     } catch (Exception e) {
-      throw new TiClientInternalException("Error Closing Store client.", e);
+      throw new TiClientInternalException("Error scanning data from region.", e);
     }
     return true;
   }
 
-  private boolean cacheDrain() {
+  private boolean isCacheDrained() {
     return currentCache == null
         || index >= currentCache.size()
         || index == -1;
   }
 
   @Override
   public boolean hasNext() {
-    if (cacheDrain() && endOfRegion) {
-      return false;
-    }
-    if (cacheDrain()) {
-      if (!loadCache()) {
-        endOfRegion = true;
-      }
-    }
-    if (!cacheDrain() && !contains(currentCache.get(index).getKey())) {
-      endOfRegion = true;
+    if (isCacheDrained() && !loadCache()) {
+      endOfScan = true;
       return false;
     }
     return true;
   }
 
   private Kvrpcpb.KvPair getCurrent() {
-    if (cacheDrain() && endOfRegion) {
-      throw new NoSuchElementException();
+    if (isCacheDrained()) {
```

Comment:
do we need endofScan?

Reply:
Just a guard not to call PD again if user call hasNext multiple times.